### PR TITLE
Adding support for Wink set target temperature async.

### PR DESCRIPTION
**Pull request by EllenPorter**

### CODE REVIEW

**Comment by EllenPorter** [comment]
Need to update mock response and remove old resourceSchemaToProviderSchema.

File: org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js [hunk 1]
```diff
@@ -279,6 +279,149 @@ function validateResourceGet(resourceId) {
...
+ * If no unit is provided and the value is within one of the two ranges,
+ *  the unit of the valid range will be returned, otherwise an 
+ *  exception is thrown.
+ *  * 
+ * @param {*} resourceSchema 
+ * @param {*} providerSchema 
+ */
+function getValidatedUnit(resourceSchema, stateReader) {
+    var value = resourceSchema.temperature;
+    if (resourceSchema.hasOwnProperty('units') && resourceSchema.units != null) {
```
Comment (rgruian) [inline org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js]: same observation here. "units" may be undefined, at least theoretically.

File: org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js [hunk 2]
```diff
@@ -279,6 +279,149 @@ function validateResourceGet(resourceId) {
...
+    var result = { 'desired_state': {} };
+    var desired_state = result.desired_state;
+
+    var value = resourceSchema.temperature;
+    var unit = getValidatedUnit(resourceSchema, stateReader);
+    if (unit === 'f') {
+        value = fahrenheitToCelsius(value);
+    }
+
+    var min_min = getMinTemperatureLow(stateReader);
```
Comment (rgruian) [inline org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js]: min_low, min_high, max_low, max_high?

File: org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js [hunk 3]
```diff
@@ -279,6 +279,149 @@ function validateResourceGet(resourceId) {
...
+function getMinTemperatureLow(stateReader) {
+    // lowest allowed max set point
+    var min_min_c = stateReader.get("min_min_set_point");
+    return min_min_c === undefined ? 9 : min_min_c;
+}
+
+function getMinSetPoint(stateReader) {
+    // min set point for heating in celsius
+    var min_c = stateReader.get("min_set_point");
+    return min_c === undefined ? 22 : min_c;
```
Comment (rgruian) [inline org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js]: min_c could theoretically be null. You may want to guard against that too, perhaps use:
((typeof min_c) == "undefined") ? 22 : min_c

Same below.

File: org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js [hunk 4]
```diff
@@ -138,14 +138,16 @@ function defaultValueIfEmpty(property, defaultValue) {
...
     var stateReader = new StateReader(providerSchema.desired_state, providerSchema.last_reading);
 
-    var max = stateReader.get('max_set_point');
-    var min = stateReader.get('min_set_point');
-
     // according to Wink docs the temperature is always returned as C 
-    var temperatureUnits = "c";
+    // state reader shows a 'units' field, return the result in this unit if it exists
+    var units = stateReader.get('units');
+    var temperatureUnits = (units === undefined || units.temperature === undefined) ? 'c' : units.temperature;
```
Comment (muralidhars) [inline org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js]: can units be null?

File: org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js [hunk 5]
```diff
@@ -279,6 +279,149 @@ function validateResourceGet(resourceId) {
...
+    desired_state['min_set_point'] = newMinSetPoint;
+    desired_state['max_set_point'] = newMaxSetPoint;
+
+    return result;
+}
+
+function getMaxTemperatureHigh(stateReader) {
+    // highest allowed max set point
+    var max_max_c = stateReader.get("max_max_set_point");
+    return max_max_c === undefined ? 32 : max_max_c;
```
Comment (muralidhars) [inline org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js]: same comments here..can max_max_c be null?

File: org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js [hunk 6]
```diff
@@ -138,14 +138,16 @@ function defaultValueIfEmpty(property, defaultValue) {
 function providerSchemaToPlatformSchema(providerSchema, expand) {
     var stateReader = new StateReader(providerSchema.desired_state, providerSchema.last_reading);
 
-    var max = stateReader.get('max_set_point');
-    var min = stateReader.get('min_set_point');
-
     // according to Wink docs the temperature is always returned as C 
-    var temperatureUnits = "c";
+    // state reader shows a 'units' field, return the result in this unit if it exists
+    var units = stateReader.get('units');
```
[the remark this answers is not in the document]
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js]: Fixed.

File: org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js (hunk 1 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js]: Fixed.

File: org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js (hunk 2 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js]: Good idea, fixed.

File: org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js (hunk 3 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js]: Fixed.

File: org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js (hunk 4 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js]: Fixed.

File: org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js (hunk 5 above)
Comment (EllenPorter) [reply org.opent2t.sample.thermostat.superpopular/com.wink.thermostat/js/thingTranslator.js]: Fixed.

**Pull request merged**